From 27d58fb0d8d978c77b3625221d45cd4a06ca201d Mon Sep 17 00:00:00 2001
From: Dong-Yuan Chen <dong-yuan.chen@intel.com>
Date: Tue, 7 May 2013 15:22:29 -0700
Subject: Dalvik: Fixed a Klocwork critical issue in printTrace

BZ: 105191

Fixed a Klocwork identified critical issue where next_code_ptr may be
used before it is initialized in printTrace.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I04ee3cd0a7928f558c7c5b9713af572372ebc630
Orig-MCG-Change-Id: Ie943a566327ec67b020296abb724aed186e9962e
Signed-off-by: Dong-Yuan Chen <dong-yuan.chen@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/CodegenInterface.cpp |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/vm/compiler/codegen/x86/CodegenInterface.cpp b/vm/compiler/codegen/x86/CodegenInterface.cpp
index 05ee268..03cee3f 100644
--- a/vm/compiler/codegen/x86/CodegenInterface.cpp
+++ b/vm/compiler/codegen/x86/CodegenInterface.cpp
@@ -1594,6 +1594,8 @@ static void printTrace(CompilationUnit *cUnit, std::vector<CodeBlockElem> &code_
     ALOGD("** %s%s@%#x:", cUnit->method->clazz->descriptor,
           cUnit->method->name, cUnit->traceDesc->trace[0].info.frag.startOffset);
 
+    code_ptr = NULL;
+    next_code_ptr = NULL;
     for (k = 0; k < (code_block_table.size() - 1); k++) {
         blk_type = code_block_table[k].first;
         code_ptr = code_block_table[k].second;
@@ -1629,6 +1631,11 @@ static void printTrace(CompilationUnit *cUnit, std::vector<CodeBlockElem> &code_
         }
     }
 
+    if (next_code_ptr == NULL) {
+        // simply return if there is no entry in code block
+        return;
+    }
+
     // next_code_ptr should hold the pre-padded address of the chaining cell count section
     // print the chaining cell count section
     next_code_ptr = reinterpret_cast<char*>((reinterpret_cast<unsigned int>(next_code_ptr) + 3) & ~0x3);
-- 
1.7.4.1

